src/ugfx.c

```diff
@@ -0,0 +1,180 @@
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include "ugfx.h"
+#include "rsp.h"
+#include "n64sys.h"
+
+#define UGFX_PARAMS_ADDRESS 0x0
+
+DEFINE_RSP_UCODE(rsp_ugfx);
+
+static void *ugfx_rdp_buffer;
+static uint32_t ugfx_rdp_buffer_size;
+
+extern void *__safe_buffer[];
+
+typedef struct ugfx_ucode_params_t {
+    uint32_t input;
+    uint32_t input_size;
+    uint32_t output_buf;
+    uint32_t output_buf_size;
+} __attribute__ ((aligned (16))) ugfx_ucode_params_t;
+
+void ugfx_matrix_from_column_major(ugfx_matrix_t *dest, const float *source)
+{
+    for (size_t i = 0; i < 16; ++i)
+    {
+        size_t c = i / 4;
+        size_t r = i % 4;
+        int32_t fixed = float_to_fixed(source[i], 16);
+        dest->integer[c][r] = (uint16_t)((fixed & 0xFFFF0000) >> 16);
+        dest->fraction[c][r] = (uint16_t)(fixed & 0x0000FFFF);
+    }
+}
+
+void ugfx_matrix_from_row_major(ugfx_matrix_t *dest, const float *source)
+{
+    for (size_t i = 0; i < 16; ++i)
+    {
+        size_t c = i % 4;
+        size_t r = i / 4;
+        int32_t fixed = float_to_fixed(source[i], 16);
+        dest->integer[c][r] = (uint16_t)((fixed & 0xFFFF0000) >> 16);
+        dest->fraction[c][r] = (uint16_t)(fixed & 0x0000FFFF);
+    }
+}
+
+ugfx_buffer_t* ugfx_buffer_new(uint32_t capacity)
+{
+    ugfx_buffer_t *buffer = malloc(sizeof(ugfx_buffer_t));
+
+    buffer->data = malloc(capacity * sizeof(ugfx_command_t));
+    buffer->capacity = capacity;
+    buffer->length = 0;
+
+    return buffer;
+}
```

Comment:
Good idea! I think this is closely related to the idea of a modular microcode system (as mentioned in another comment). So instead of a display list that is managed by ugfx, it would be a more generic component.

I think by making the user manage the command buffers I wanted to avoid having a singleton-like API, but in hindsight that doesn't really make sense in the context of libdragon. Maybe just a force of habit 😄